x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/slm/SnapshotRetentionTask.java

```diff
@@ -247,21 +256,34 @@ void deleteSnapshots(Map<String, List<SnapshotInfo>> snapshotsToDelete,
 
         logger.info("starting snapshot retention deletion for [{}] snapshots", count);
         long startTime = nowNanoSupplier.getAsLong();
-        int deleted = 0;
+        final AtomicInteger deleted =  new AtomicInteger(0);
+        final AtomicInteger failed = new AtomicInteger(0);
         for (Map.Entry<String, List<SnapshotInfo>> entry : snapshotsToDelete.entrySet()) {
             String repo = entry.getKey();
             List<SnapshotInfo> snapshots = entry.getValue();
             for (SnapshotInfo info : snapshots) {
-                deleteSnapshot(getPolicyId(info), repo, info.snapshotId(), slmStats);
-                deleted++;
+                deleteSnapshot(getPolicyId(info), repo, info, slmStats, historyItem -> {
+                    // This would be nicer if we could use ifPresentOrElse
+                    historyItem.ifPresent(item -> {
```

Comment:
Hmm... I wonder if rather than presenting an Optional and using it (or null) as an indicator of a failure, this should conform to the `ActionListener` interface, and then it could be

```java
ActionListener.wrap(item -> { deleted.incrementAndGet(); historyStore.putAsync(item); }, e -> failed.incrementAndGet())
```

That would prevent having to do the `ifSuccess` and `ifEmpty` checks, as well as not pass an `Optional` around